Adding a basic service for loading and using .NET

tannergooding opened the pull request:
This adds a basic service for loading and using .NET. This uses the official hosting APIs and therefore works with normal redistribution of the .NET Runtime.

tannergooding commented:
![image](https://user-images.githubusercontent.com/10487869/65337867-dad07400-db7d-11e9-8e6e-57176f687fa1.png)

tannergooding commented on src/NovelCoreClrService.cpp:
```diff
@@ -0,0 +1,130 @@
+const auto HostApiBufferTooSmall = 0x80008098;
+
+static void* getNativeExport(void* nativeLibrary, const char* exportName) {
+#if defined(_WIN32)
```
Need to have some logic for loading and finding the hostfxr exports that differs between Windows and Unix.

tannergooding commented on src/NovelCoreClrService.cpp:
```diff
@@ -0,0 +1,130 @@
+
+namespace NovelRT {
+  NovelCoreClrService::NovelCoreClrService() :
+    _hostContextHandle(Lazy<hostfxr_handle>([&, this] {
```
All the function pointers, the hostfxr resolution logic, etc are lazy.

tannergooding commented on src/NovelCoreClrService.h:
```diff
@@ -0,0 +1,54 @@
+  };
+
+  template <class _Fty>
+  std::function<_Fty> NovelCoreClrService::getFunction(const char_t* assemblyPath, const char_t* typeName, const char_t* methodName, const char_t* delegateTypeName) {
```
This could be made a bit "simpler" and have some strings be auto-constructed, etc.
However, this is zero overhead and most things (like the Ink service) will be their own APIs that can abstract this away.

tannergooding commented on src/main.cpp:
```diff
@@ -80,6 +80,9 @@ int main(int argc, char* argv[]) {
   rect->subscribeToInteracted([]{novelChanRect->setActive(!novelChanRect->getActive());});
 
+  auto coreclrService = runner.getCoreClrService();
+  auto testFunction = coreclrService->getFunction<void()>(STR("ClassLibrary1.dll"), STR("ClassLibrary1.Class1, ClassLibrary1"), STR("Test"), STR("ClassLibrary1.TestSignature, ClassLibrary1"));
```
ClassLibrary1.dll isn't included here, since it was just a test.

Before the PR is merged, we should determine how cmake will pull down any managed NuGet packages that are dependencies and replace this with using that instead...

tannergooding replied:
This was changed to a new `NovelRT.Dotnet` csproj which is checked in, but which requires a bit more infrastructure work to properly integrate with the build system.

RubyNova replied:
Just to flag this up, we should probably decide together what's handled by the SDK vs what's handled by the engine core in this aspect.

Specifically, should the SDK be handling `dotnet restore` instead as a whole, for example? Or did you want to split it out so that it only handles `dotnet restore` for end-user csproj files?

tannergooding commented:
Test

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/novelrt/NovelRT?pullRequest=54) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/novelrt/NovelRT?pullRequest=54) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/novelrt/NovelRT?pullRequest=54) it.</sub>

tannergooding commented:
Updated this to use the official vcpkg and renamed a few components. The unresolved comments from above are still valid and apply to the corresponding locations in the renamed files.

RubyNova commented on include/NovelRT.h:
```diff
@@ -120,6 +128,7 @@ namespace NovelRT::Windowing {
 //Engine service types
 #include "NovelRT/Audio/AudioService.h"
 #include "NovelRT/DebugService.h"
+#include "NovelRT/Dotnet/DotnetRuntimeService.h"
```
This is just a question really about naming - since its already in `NovelRT::Dotnet` would `RuntimeService` be a better type name? or perhaps `ClrService`? Maybe this is just me, but were calling it `Dotnet` twice here.

I'm also possibly wondering about renaming the namespace to `DotNet` as opposed to `Dotnet` maybe? What do you think?

tannergooding replied:
Renamed

RubyNova commented on include/NovelRT/DotNet/RuntimeService.h:
```diff
@@ -0,0 +1,51 @@
+namespace NovelRT::DotNet {
+  class RuntimeService {
+  private:
+    NovelRT::Utilities::Lazy<hostfxr_handle> _hostContextHandle;
```
Base namespace is the same; you could just use `Utilities::Lazy<T>` instead. This applies to the others too. 😄

RubyNova commented on src/CMakeLists.txt:
```diff
@@ -32,25 +25,36 @@ set(NOVELRT_LIB_LINK_LIBRARIES
+link_libraries(${NETHOST_LIBRARY})
 
 if (WIN32)
   set(CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS ON)
```
Is there a reason this is WIN32 specific?

tannergooding replied:
Linux automatically does the right thing.

RubyNova replied:
Ah oki doki. Resolving.

RubyNova commented on src/NovelRT/DotNet/RuntimeService.cpp:
```diff
@@ -0,0 +1,148 @@
+
+#include <NovelRT.h>
+
+#include <nethost.h>
```
Uhhhh is there a reason this isn't just the singular `NovelRT.h` include?

tannergooding replied:
It's a header only needed by the implementation and is not needed for 3rd party libraries to also include.

RubyNova replied:
Ah okay, awesome.

RubyNova commented on src/NovelRT/DotNet/RuntimeService.cpp:
```diff
@@ -0,0 +1,148 @@
+  if (result != 0)
+  {
+    std::cerr << "ERROR: failed to locate hostfxr" << std::endl;
+    throw - 1;
```
We should probably throw `std::runtime_error` here in the same vein as the other places (even if its not what we want to do in the future, it'd be nice to have this uniform), unless this is going to be exported as a C function or something?

tannergooding replied:
Ah right, forgot to fix these now that we have a logger.

RubyNova commented on src/NovelRT/DotNet/RuntimeService.cpp:
```diff
@@ -0,0 +1,148 @@
+
+  if (result != 0)
+  {
+    std::cerr << "ERROR: failed to locate hostfxr" << std::endl;
```
We should probably be using our logger type here.

RubyNova commented on src/NovelRT/DotNet/RuntimeService.cpp:
```diff
@@ -0,0 +1,148 @@
+
+      if (result != 0)
+      {
+        std::cerr << "ERROR: failed to initialize the runtime" << std::endl;
```
I won't repeat myself on these next two lines of code, but basically the same suggestions as lines 76 and 77. 😄
